File: src/house_robber_198/solution_recursion.ts
function maxRob(nums: number[], cache: Map<number, number>, idx: number): number {
  if (idx >= nums.length) {
    return 0;
  }

  if (cache.has(idx)) {
    return cache.get(idx)!;
  }

  const result = Math.max(nums[idx] + maxRob(nums, cache, idx + 2), maxRob(nums, cache, idx + 1));

  cache.set(idx, result);
  return result;
}

function rob(nums: number[]): number {
  const cache = new Map<number, number>();
  return maxRob(nums, cache, 0);
}

export function test(): void {
  console.log(rob([1, 2, 3, 1]));
  console.log(rob([2, 7, 9, 3, 1]));
}

File: src/house_robber_198/solution_recursion.py
from typing import List, Dict


class Solution:
    def max_rob(self, nums: List[int], cache: Dict[int, int], idx: int) -> int:
        if idx >= len(nums):
            return 0

        if idx in cache:
            return cache[idx]

        result = max(
            nums[idx] + self.max_rob(nums, cache, idx + 2),
            self.max_rob(nums, cache, idx + 1),
        )

        cache[idx] = result
        return result

    def rob(self, nums: List[int]) -> int:
        cache: Dict[int, int] = {}
        return self.max_rob(nums, cache, 0)


def test():
    solution = Solution()
    print(solution.rob([1, 2, 3, 1]))
    print(solution.rob([2, 7, 9, 3, 1]))
